[Concurrency][Stdlib] Add SwiftStdlibCurrentOS availability, use it.

al45tair opened the pull request:
If you use `SwiftStdlibCurrentOS` availability, you will be able to use new types and functions from within the implementation. This works by, when appropriate, building with the `CurrentOS` availability set to the current deployment target.

rdar://150944675

al45tair committed:
[Concurrency][Stdlib] Add SwiftStdlibCurrentOS availability, use it.

If you use SwiftStdlibCurrentOS availability, you will be able to
use new types and functions from within the implementation. This
works by, when appropriate, building with the CurrentOS availability
set to the current deployment target.

rdar://150944675

al45tair commented:
@swift-ci Please test

stephentyrone commented:
Love where this is going, but I think the names are more-or-less backwards; we shouldn't reference the "OS" for changes that are just new API additions in the stdlib at all. They should be "SwiftStdlib x.y," and the rare change that _does_ depend on an outside OS feature should either reference the OS version, or reference the feature itself (Foundation does this).

etcwilde commented on Runtimes/Core/cmake/modules/AvailabilityMacros.cmake:
```diff
@@ -5,4 +5,22 @@ file(STRINGS "${CMAKE_CURRENT_BINARY_DIR}/availability-macros.def" availability_
+    # ON will cause us to use the "proper" availability instead.
+    string(REPLACE "SwiftStdlib" "SwiftStdlibCurrentOS" current "${def}")
+    if(NOT SwiftCore_ENABLE_STRICT_AVAILABILITY AND SwiftCore_SWIFT_AVAILABILITY_PLATFORM)
+      string(REGEX MATCH "${SwiftCore_SWIFT_AVAILABILITY_PLATFORM} ([0-9]+(\.[0-9]+)+)" platform_version "${def}")
```
AvailabilityMacros.cmake is loaded before platform info, so `SwiftCore_SWIFT_AVAILABILITY_PLATFORM` won't be set by that here.

al45tair commented:
> Love where this is going, but I think the names are more-or-less backwards

Agreed, _but_ I didn't want to change the meaning of `SwiftStdlib`; for context, the reason I'm doing this _now_ is that when we tell Swift what `SwiftStdlib 6.2` is, a load of Concurrency things will break because Concurrency wants to use new things to implement other things, but that won't work on the CI machines (or in OSS) where the build system is a lower version than whatever `SwiftStdlib 6.2` ends up being.

We could make changing the names a separate task, if someone comes up with something better.

al45tair committed:
[Build] Add catalyst support to SwiftStdlibCurrentOS.

Add catalyst support for SwiftStdlibCurrentOS.

Also, set a minimum deployment target when building Concurrency;
this stops the build failing when we're trying to build on older
systems where Concurrency didn't really exist yet.

rdar://150966361

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test macOS platform

al45tair committed:
[Build] Fix variable name in new build system.

It should have been `SwiftCore_MODULE_TRIPLE`, not
`SwiftCore_SWIFT_MODULE_TRIPLE`.

al45tair committed:
[Build] Use `FATAL_ERROR`, not `ERROR`.

We should be using `FATAL_ERROR` for error messages from CMake, not
just `ERROR`.

al45tair committed:
[Build] Fix macos vs macosx issue.

The module triple bizarrely omits the "x" from "macosx".

Also disable some of this code on non-Apple platforms.

rdar://150944675

al45tair commented:
@swift-ci Please smoke test

stephentyrone commented:
I don't think you need to fully solve the naming issue now, but _please_ do not merge this under a name that references "OS"; there will be considerable inertia to changing this later, so we should not use a name that we know is bad. If we need a placeholder name, let's call this `@available(DeploymentTarget)` or similar, which is at least not _wrong_.

al45tair committed:
[Build] Fix availability platform macros.

These need to be defined always; we'll set them to "unknown" and "none"
respectively if they end up being something we don't understand.

rdar://150966361

al45tair commented:
@swift-ci Please smoke test

al45tair commented:
@swift-ci Please smoke test
